### ISSUE TITLE
Future plans

### ISSUE THREAD

**Issue by seanegoodwin** [seanegoodwin/PandaViewer#78]
Just thought I'd make a statement on where I think PV will be going in the next few months/arbitrary timeframe.

I'm sure anyone who has paid attention to the project noticed a more or less complete stop in development last September. Between my job, finishing my degree, and a variety of other things I simply haven't mad much time to devote to the project. Hopefully things will be wrapping up in the next two months or so and I can begin again.

An additional factor for the lack of progress is the general architecture/code quality of the project, or the lack thereof. While recent changes and additions tend to be of a much higher quality, a lot of the old stuff simply sucks. When you consider the history of Pandaviewer, from a simple collection of stateless scripts that scanned your collection on startup to now, it makes some sense. 

A lot of the features I want to implement would be difficult without a major refactor effort, which goes back to the lack of free time. I've started some initial planning, but nothing concrete. I hope to ramp up work on this sometime in June, but time frames are malleable. 

Anyway, for now please keep posting suggestions and bug reports. I read them all, even if I don't respond for a long time.

**Comment by Black-Light** [comment]
Are you still alive? PandaViewer still works very well but I hope you fix some of the important issues such as #70 and #76 (or an option/temporary removal of the autodetection feature since it causes more harm then good currently) Thanks for all the hard work you put into this project.

**Comment by seanegoodwin** [comment]
Definitely not dead (yet). I've been meaning to make another post on the present/future of the project and I guess now is as good as any time. I'll post a TLDR at the end for people not interested in blogshit.

As anyone who has been paying attention would have noticed, development on PV has been basically nonexistent for around a year now (one minor update early this year aside).
There are a number of reasons for this, mostly personal, some technical. The project as it stands ended up ballooning in scope from when I started it in June of 2014. At the time, it was basically a 300 line program that would scan your directories for collections of images on startup and display the first image of each and allow you to open the file by clicking it. That was all. There was no database, no metadata collection, no duplicate finder/detection, no directory watcher, no fancy GUI, little searching, no archive support, etc, etc. You get the idea.

And while the structure of PV has changed a lot since those days, a lot of technical debt and poor decisions have led the project to it's current state. There are a lot of problems with the design of the program that really inhibit development. I won't get to far into details, but I'll list some of the terrible decisions that lead to this:
- Over reliance on state. Because the program wasn't really intended to work with a database, and the database implementation was hacked on, the program basically requires all galleries to be constantly loaded into memory. While the memory usage is trivial, the problem is that we start running into situations where I have the reconcile the state between the database, the gallery object, and the actual, file(s) the gallery represents. This is a massive headache and a source of a large amount of dev time.
- Listening to users too much. This might be relatively minor, but I figure it's worth mentioning.
    I love getting bug reports and suggestions, but I've historically been too eager to appease you guys and add features that aren't really supported well, or are a pain to maintain. This is entirely a personal failing where I didn't really know when to say no or to take the required time to think about the implications of a feature. The directory watching is a prime example of this.
- Not listening to users enough. As a counterpoint to the previous, there were a number of times where I spent a massive amount of hours developing features that were ultimately unwanted. Proper user discussions and polling should be required before starting work on any major feature.
- Poor planning. I touched on this a bit earlier, but PV had a really rough time growing up. There was very little thought that went into it's early development, mostly because I was the only one using it for around 8 months or so. I wasn't experienced in releasing software to what eventually became hundreds or thousands of people, nor was I prepared for the effort that would have to be spent debugging issues.
    
Additionally, I haven't had the time or desire to work on it for a while. When I spend 8+ hours at work programming it's pretty much the last thing I want to do when I get home.
I started on a redesign of the core backend sometime around March or April in another Github repo, but I haven't made much progress beyond prototyping and tech testing. The implementation is exciting and I'm trying to slowly get back into development.
Some of the features of the new implementation would include:
- Massively improved collection management. You would be able to not only view individual galleries like the current implementation, but view/create series/collections of galleries. The reverse would also be possible: given a tank, you could create a release out of it and split the individual galleries through the program.
- Passive duplicate detections. The new DB structure would allow PV to detect any duplicate files as a matter of fact and without any additional scanning/hashing.
- Web interface(s). By structuring the backend as a simple REST server, you have the additional benefit of being able to potentially access your collection anywhere. You could run PV on your home computer and access it from any other computer/device on the network through a web interface. Think Plex for manga/doujinshi.
- Better interface. QML sucks. No one uses it. It's really hard to find documentation on certain things and there are much fewer resources available then I would like.
- Proper searching. Searching right now sucks, both from the user perspective and the development perspective.
- A bunch of other stuff I'm too lazy to list right now.

But ultimately, I want to see if there is interest in these things or a future version of PV. There are a lot of alternatives out there now, and the ecosystem has changed a lot in two years. I see a lot of people Hydrus and similar management tools in the threads nowadays, as apposed to PV (or HP for that matter). Is there still room for PV in the grand scheme of things? I have no idea.

TLDR:
I am trying to get back into development, but it will likely be focused on a new, drastically different implementation. One I wouldn't mind supporting in the long term.
There is potential room for a finalish patch for the current version of PV to address some the easier to fix issues, but new features are unlikely on the current architecture.
Expect me to do some polling over the next couple of weeks to try and gather interest in specific features/a new version of PV at all. 
I'd love to get feedback on what people's current usage of PV (or similar tools if that's what you use), what features are used/not used, what features are really wanted, etc.

**Comment by Black-Light** [comment]
Thanks for the update, it's been pretty amazing watching the development of this project, it certainly made managing my mess of an archive manageable. You don't need to apologize for taking a long time in between updates, after all your working on this in your free time. I'm glad you want to start having a bit more dialog with the PV users for upcoming features, spending lots of time on something only a few people will use isn't worth it (unless it's a feature you want yourself). If anything is too hard to implement you need to just say no, people can always make a pull request.

The big things for me are searching and sorting methods, Tag generation (not much to be improved here) and managing tanks and all three of these were addressed in your post. I only use PV at the moment I have tried some other software but PV has always been the easiest and quickest for me to use.

In terms of usage I spend hours tagging galleries with PV for the sake of archiving. Managing these huge unsorted folders in PV saved me from spending months renaming folders (not that i would of done that). Anything that could make tagging faster would make me very happy such as a small QoL feature to easily paste metadata links on galleries by right-click "paste meta" them instead of: click options > click on metadata field > paste > click save > click back.

Features like that are probably not too hard to implement but would save users lots of time, If your not prepared to spend hours and hours (understandably) on this project then having a dialog with us and managing your time better would benefit the project overall. The web interface is a great idea for this kind of tool, However I don't know how necessary it is considering most of these galleries are currently available online. Would be nice to see how many people would be interested in this feature.

Thanks for the update looking forward to trying the new versions whenever they become available.

P.S Sorry for requesting yet another feature, in this post..

**Comment by seanegoodwin** [comment]
Just like to post a small update. I've now resumed work on the new version of PandaViewer. I don't have any idea of when either a beta/more information may be available, but I will be working on it again.
